### PR TITLE
Enhance ReviewUnusedParameter to bail out if $MyInvocation.BoundParameters encountered

### PR DESCRIPTION
**Pull request by jegannathanmaniganadan**
## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [ ] [Make sure all `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] Make sure you've added a new test if existing tests do not effectively test the code changed and/or updated documentation
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

### CODE REVIEW

File: Rules/ReviewUnusedParameter.cs [hunk 1]
```diff
@@ -51,6 +51,29 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
...
                 }
 
+                // bail out if $MyInvocation.BoundParameters used
+                if (variableCount.ContainsKey("MyInvocation"))
+                {
+                    bool isBound = false;
+                    IEnumerable<Ast> memExpAst = scriptBlockAst.FindAll(oneAst => oneAst is MemberExpressionAst, false);
+                    foreach (MemberExpressionAst memberExpressionAst in memExpAst)
+                    {                        
+                        if (memberExpressionAst.Expression is VariableExpressionAst &&
```
Comment (mattpwhite) [inline Rules/ReviewUnusedParameter.cs]: Use the nicer pattern matching syntax that allows you to conditionally create a new typed local as part of the `is` condition - https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7#pattern-matching.

```
memberExpressionAst.Expression is VariableExpressionAst veAst &&
veAst.VariablePath.UserPath...
```

File: Rules/ReviewUnusedParameter.cs [hunk 2]
```diff
@@ -51,6 +51,29 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
...
+                            break;
+                        }                        
+                    }
+
+                    if (isBound)
+                    {
+                        continue;
+                    }
+                }
+
```
Comment (mattpwhite) [inline Rules/ReviewUnusedParameter.cs]: What about `$PSCmdlet.MyInvocation.BoundParameters`?

File: Rules/ReviewUnusedParameter.cs (hunk 1 above)
Comment (jegannathanmaniganadan) [reply Rules/ReviewUnusedParameter.cs]: Nice, thanks for the suggestion

File: Rules/ReviewUnusedParameter.cs (hunk 2 above)
Comment (jegannathanmaniganadan) [reply Rules/ReviewUnusedParameter.cs]: Good catch, let me enhance this condition to handle this as well.

Comment (jegannathanmaniganadan) [reply Rules/ReviewUnusedParameter.cs]: I have tweaked this for `$PSCmdlet.MyInvocation.BoundParameters` as well, do you think it can be optimized further or is this okay

Comment (jegannathanmaniganadan) [reply Rules/ReviewUnusedParameter.cs]: I have different approach, will implement it and let you know

File: Rules/ReviewUnusedParameter.cs [hunk 3]
```diff
@@ -36,6 +36,12 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
 
             foreach (ScriptBlockAst scriptBlockAst in scriptBlockAsts)
             {
+                // bail out if PS bound parameter used.
+                if (scriptBlockAst.FindAll(IsBoundParametersReference, false).Any())
```
Comment (mattpwhite) [inline Rules/ReviewUnusedParameter.cs]: Would `Find() != null` be better than `FindAll().Any()`?

File: Rules/ReviewUnusedParameter.cs [hunk 4]
```diff
@@ -72,6 +72,34 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
             }
         }
 
+        /// <summary>
+        /// Checks for PS bound parameter reference.
+        /// </summary>
+        /// <param name="ast"></param>
```
Comment (mattpwhite) [inline Rules/ReviewUnusedParameter.cs]: Either fill in this param description and returns or remove them

File: Rules/ReviewUnusedParameter.cs [hunk 5]
```diff
@@ -72,6 +72,34 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
...
+                    && veAst.VariablePath.UserPath.Equals("MyInvocation", StringComparison.OrdinalIgnoreCase))
+                    || (memberAst.Expression is MemberExpressionAst meAst
+                    && meAst.Expression is VariableExpressionAst veAstNested
+                    && veAstNested.VariablePath.UserPath.Equals("PSCmdlet", StringComparison.OrdinalIgnoreCase)
+                    && meAst.Member is StringConstantExpressionAst sceAstNested
+                    && sceAstNested.Value.Equals("MyInvocation", StringComparison.OrdinalIgnoreCase));
+            }
+
+            return isBound
+                || ast is VariableExpressionAst variableAst
```
Comment (mattpwhite) [inline Rules/ReviewUnusedParameter.cs]: This whole method is pretty hard to follow.  I would recommend re-writing it to be a clearer set of conditions.

```
        private static bool IsBoundParametersReference(Ast ast)
        {
            if (!(ast is VariableExpressionAst variableAst) || 
                variableAst.VariablePath.UserPath.Equals("PSBoundParameters", StringComparison.OrdinalIgnoreCase))
            {
                 return false;
            }

            // and so on - with ifs using no more than 3 conditions.

        }
```

**Review by mattpwhite — CHANGES_REQUESTED** [review]

File: Rules/ReviewUnusedParameter.cs (hunk 3 above)
Comment (jegannathanmaniganadan) [reply Rules/ReviewUnusedParameter.cs]: Yes, it think since it returns the first matching node. Cost effective

File: Rules/ReviewUnusedParameter.cs (hunk 5 above)
Comment (jegannathanmaniganadan) [reply Rules/ReviewUnusedParameter.cs]: I hope 3 condition is not a hard limit? I updated this function to have readable if conditions. The condition for checking $PSCmdlet.MyInvocation still has more than 3 conditions (i.e 4).
